_posts/2021-07-13-ci-metrics.md

```diff
@@ -0,0 +1,116 @@
+
+
+ 1. A merged PR becomes fully green with a 75% chance at the first attempt, and with a 95% chance after one retry.
+ 2. Every individual test succeeds at least 90% of the time.
+ 3. 95% of all PRs are merged without failed tests.
+ 4. 95% of test runs take no more than 1 hour to execute.
+ 5. 95% of test runs spend no more than 5 minutes in the queue until they get assigned to a runner.
+ 6. 95% of scheduled tests run through to completion (all tests ran and status got reported to PR).
+
+ These levels are not very ambitious compared to [five nines](https://en.wikipedia.org/wiki/High_availability#Percentage_calculation) that are often required in other places. But first, we intentionally started low to get a grasp of how an error budget works; and second, when we started this we were rather far away from even meeting these. These SLOs represented a challenge which was achievable with a realistic amount of effort, while being good enough to not hamper our Cockpit development life.
+
+ We [published these SLOs on our wiki](https://github.com/cockpit-project/cockpit/wiki/DevelopmentPrinciples#our-testsci-error-budget) to generate a sense of commitment and transparency.
+
+
+## Implement Service Level Indicators
+
+Now comes the time to meet the hard reality -- to pry these numbers out of production CI. Fortunately, most of the above SLIs (all but 2.) can be derived directly from the [GitHub Statuses](https://docs.github.com/en/rest/reference/repos#statuses) history. For example, 5. is the time between "Not yet tested" and "Testing in progress" status changes, while 4. is the time between "Testing in progress" and any final (success, failure, error) status.
+
+Only 2. requires scanning the actual test logs; our logs are in [TAP format](https://testanything.org/tap-specification.html), which makes it easy to scan for lines starting with `ok` or `not ok` and reading the test name and retry attempt/reason after that.
+
+This evaluation happens in the [store-tests bots script](https://github.com/cockpit-project/bots/blob/main/store-tests). This gets called through a [webhook](https://docs.github.com/en/developers/webhooks-and-events/webhooks) for each *merged* PR. We deliberately do *not* count failures or long tests from intermediate force-pushes or abandoned PRs, as during that phase pretty much anything can (and does) go wrong. However, GitHub statuses do record test retries (for SLO 1.).
+
+`store-tests` puts its results into an [SQLite database](https://sqlite.org/index.html) which gets [published on our logs server](https://logs-https-frontdoor.apps.ocp.ci.centos.org/test-results.db).
```

Comment:
Perhaps linking to the source of the store-tests script would help the reader?

Reply:
The previous paragraph already does it, but can't hurt to repeat it. Done.